### PR TITLE
[WIP] Refactor cadvisor metric collection

### PR DESCRIPTION
**Pull request by ehashman**
This is a refactor of https://github.com/kubernetes/kubernetes/pull/106334 on latest master. In its current state it does not build since I just cherry-picked the top commit from my last PR, but the vendor update was 80% of the work so wanted to get this out here in case anyone wanted to play with it :)

/cc @dgrisonnet @bwplotka

#### What type of PR is this?

/kind feature
/sig node
/sig instrumentation

#### What this PR does / why we need it:

Significantly improves perf of cadvisor metric collection

Note: neither the NewBlockingRegistry nor the cadvisor changes have been released yet, hence the funky version pins for testing.

#### Which issue(s) this PR fixes:

Fixes #

#### Special notes for your reviewer:

I assume this will make the linter very unhappy

#### Does this PR introduce a user-facing change?

```release-note
cAdvisor uses a new type of registry for container metric collection. Users should see significant improvements in kubelet CPU utilization as a result.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -62,11 +62,11 @@ require (
 	github.com/mvdan/xurls v1.1.0
 	github.com/onsi/ginkgo v1.14.0
 	github.com/onsi/gomega v1.10.1
-	github.com/opencontainers/runc v1.0.3
+	github.com/opencontainers/runc v1.1.0
```
Comment (ehashman) [inline go.mod:65]: I am not actually sure why this is getting bumped...

/shrug

Comment (ehashman) [reply go.mod:65]: runc was bumped in cadvisor which is why this was getting picked up from this PR. Rebased on top of @kolyshkin's runc 1.1.0 patch in order to test.

**Comment by ehashman** [comment]
@dgrisonnet can you PTAL at this and see if there's any reason why metric responses would come back empty? I'm not seeing anything in the kubelet logs for cadvisor...

```
ehashman@fedora:~$ curl -k --cert /var/run/kubernetes/client-admin.crt --key /var/run/kubernetes/client-admin.key https://127.0.0.1:10250/metrics/cadvisor
ehashman@fedora:~$ less /tmp/kubelet.log 
...
I0210 15:46:57.946073  121888 httplog.go:131] "HTTP" verb="GET" URI="/metrics/cadvisor" latency="7.764438ms" userAgent="curl/7.74.0" audit-ID="" srcIP="127.0.0.1:51366" resp=0
```

**Comment by dgrisonnet** [comment]
Nothing seems wrong with the integration of the cadvisor change. I'll try to test this PR and the cadvisor one locally to see if I find anything

**Comment by bwplotka** [comment]
I managed to create a minimal local setup to test and macro benchmark this. Tried with vanilla `kubelet` and it works, will try with the code tomorrow.

See https://twitter.com/bwplotka/status/1493682298016964611

**Comment by bwplotka** [comment]
One thing I see is that this PR is crazy big. I think we should try to do the change step by step. Refactoring cadvisor usage in one go, then replace to more efficient cadvisor e.g. I will try to do the latter in separate PR.

**Comment by bwplotka** [comment]
Can repro problem, will dig further:

image

**Pull request closed without merging**